### PR TITLE
Make subfield pruning work with CBO

### PR DESCRIPTION
**Pull request by mbasmanova**
Fix the issue discussed in #13082 

```
== RELEASE NOTES ==
SPI Changes
-------------
* Add columnHandles parameter to `ConnectorMetadata.getTableStatistics` method. 
The new parameter allows connectors to implement efficient pruning of statistics to the 
desired list of columns and subfields and fixes compatibility issue between subfield pruning 
and CBO (#13082).
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 1]
```diff
@@ -944,7 +944,7 @@ private static OptionalDouble getAverageColumnLength(OptionalLong totalSizeInByt
             return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, TOTAL_SIZE_IN_BYTES, MAX_VALUE_SIZE_IN_BYTES);
         }
         if (type instanceof ArrayType || type instanceof RowType || type instanceof MapType) {
-            return ImmutableSet.of();
+            return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, TOTAL_SIZE_IN_BYTES);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java]: Could you please extract it into a separate commit? Also could you please add the tests to the `TestHiveIntegrationSmokeTest` (cases in `testCollectColumnStatisticsOnCreateTable`, `testCollectColumnStatisticsOnInsert`, `testAnalyzePartitionedTable`, `testAnalyzeUnpartitionedTable`)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java [hunk 2]
```diff
@@ -88,7 +95,7 @@
 {
     public TestHiveLogicalPlanner()
     {
-        super(() -> createQueryRunner(ImmutableList.of(LINE_ITEM), ImmutableMap.of("experimental.pushdown-subfields-enabled", "true"), Optional.empty()));
+        super(() -> createQueryRunner(ImmutableList.of(ORDERS, LINE_ITEM), ImmutableMap.of("experimental.pushdown-subfields-enabled", "true"), Optional.empty()));
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: Do we have a planner test in the `presto-hive` module? :open_mouth:

File: presto-main/src/main/java/com/facebook/presto/cost/TableScanStatsRule.java [hunk 3]
```diff
@@ -61,7 +62,10 @@ public TableScanStatsRule(Metadata metadata, StatsNormalizer normalizer)
         Map<VariableReferenceExpression, VariableStatsEstimate> outputVariableStats = new HashMap<>();
 
         for (Map.Entry<VariableReferenceExpression, ColumnHandle> entry : node.getAssignments().entrySet()) {
-            Optional<ColumnStatistics> columnStatistics = Optional.ofNullable(tableStatistics.getColumnStatistics().get(entry.getValue()));
+            // TODO Adjust statistics by taking into account subfield pruning possibly by changing
+            //  getTableStatistics SPI to pass in ColumnHandles for the columns listed in TableScanNode
+            ColumnHandle columnHandle = entry.getValue().withRequiredSubfields(ImmutableList.of());
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/cost/TableScanStatsRule.java]: This makes some implicit assumptions about the `getTableStatistics` implementation. It feels like the list of column handles has to be passed to the `getTableStatistics`, and the connector itself should assign statistics for each of the handles passed.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java (hunk 2 above)
Comment (mbasmanova) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java]: I added it to allow testing of filter pushdown and subfield pruning.

File: presto-main/src/main/java/com/facebook/presto/cost/TableScanStatsRule.java (hunk 3 above)
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/cost/TableScanStatsRule.java]: That's right. Would you like me to make that change in this PR?

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/TableScanStatsRule.java]: Yeah, i think it would be reasonable.

**Comment by mbasmanova** [comment]
@arhimondr Andrii, thank you for review. I addressed the comments and updated the PR. Would you take another look?
